File: test/language/choice_types/cpp98/DefaultEmptyChoiceTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "choice_types/default_empty_choice/DefaultEmptyChoice.h"

namespace choice_types
{
namespace default_empty_choice
{

class DefaultEmptyChoiceTest : public ::testing::Test
{
protected:
    void writeDefaultEmptyChoiceToByteArray(zserio::BitStreamWriter& writer, uint8_t tag, int32_t value)
    {
        switch (tag)
        {
        case 1:
            writer.writeSignedBits(value, 8);
            break;

        case 2:
            writer.writeSignedBits(value, 16);
            break;

        default:
            break;
        }
    }

    static const uint8_t VARIANT_A_SELECTOR;
    static const uint8_t VARIANT_B_SELECTOR;
    static const uint8_t DEFAULT_SELECTOR;
};

const uint8_t DefaultEmptyChoiceTest::VARIANT_A_SELECTOR = 1;
const uint8_t DefaultEmptyChoiceTest::VARIANT_B_SELECTOR = 2;
const uint8_t DefaultEmptyChoiceTest::DEFAULT_SELECTOR = 3;

TEST_F(DefaultEmptyChoiceTest, bitStreamReaderConstructor)
{
    const uint8_t tag = VARIANT_A_SELECTOR;
    zserio::BitStreamWriter writer;
    const int8_t value = 99;
    writeDefaultEmptyChoiceToByteArray(writer, tag, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);

    const DefaultEmptyChoice defaultEmptyChoice(reader, tag);
    ASSERT_EQ(tag, defaultEmptyChoice.getTag());
    ASSERT_EQ(value, defaultEmptyChoice.getA());
}

TEST_F(DefaultEmptyChoiceTest, read)
{
    const uint8_t tag = VARIANT_A_SELECTOR;
    zserio::BitStreamWriter writer;
    const int8_t value = 99;
    writeDefaultEmptyChoiceToByteArray(writer, tag, value);
    size_t writeBufferByteSize;
    const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
    zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
    DefaultEmptyChoice defaultEmptyChoice;
    defaultEmptyChoice.initialize(tag);
    defaultEmptyChoice.read(reader);

    ASSERT_EQ(tag, defaultEmptyChoice.getTag());
    ASSERT_EQ(value, defaultEmptyChoice.getA());
}

TEST_F(DefaultEmptyChoiceTest, write)
{
    const uint8_t tagA = VARIANT_A_SELECTOR;
    DefaultEmptyChoice defaultEmptyChoice;
    defaultEmptyChoice.initialize(tagA);

    const VariantA valueA = 99;
    defaultEmptyChoice.setA(valueA);
    zserio::BitStreamWriter writerA;
    defaultEmptyChoice.write(writerA);
    size_t writeBufferByteSizeA;
    const uint8_t* writeBufferA = writerA.getWriteBuffer(writeBufferByteSizeA);
    zserio::BitStreamReader readerA(writeBufferA, writeBufferByteSizeA);
    DefaultEmptyChoice readDefaultEmptyChoiceA(readerA, tagA);
    ASSERT_EQ(valueA, readDefaultEmptyChoiceA.getA());

    const uint8_t tagB = VARIANT_B_SELECTOR;
    defaultEmptyChoice.initialize(tagB);
    const VariantB valueB = 234;
    defaultEmptyChoice.setB(valueB);
    zserio::BitStreamWriter writerB;
    defaultEmptyChoice.write(writerB);
    size_t writeBufferByteSizeB;
    const uint8_t* writeBufferB = writerB.getWriteBuffer(writeBufferByteSizeB);
    zserio::BitStreamReader readerB(writeBufferB, writeBufferByteSizeB);
    DefaultEmptyChoice readDefaultEmptyChoiceB(readerB, tagB);
    ASSERT_EQ(valueB, readDefaultEmptyChoiceB.getB());

    const uint8_t tagDefault= DEFAULT_SELECTOR;
    defaultEmptyChoice.initialize(tagDefault);
    zserio::BitStreamWriter writerDefault;
    defaultEmptyChoice.write(writerDefault);
    size_t writeBufferByteSizeDefault;
    const uint8_t* writeBufferDefault = writerDefault.getWriteBuffer(writeBufferByteSizeDefault);
    zserio::BitStreamReader readerDefault(writeBufferDefault, writeBufferByteSizeDefault);
    DefaultEmptyChoice readDefaultEmptyChoiceDefault(readerDefault, tagDefault);
    ASSERT_EQ(tagDefault, readDefaultEmptyChoiceDefault.getTag());
}

} // namespace default_empty_choice
} // namespace choice_types

File: test/language/choice_types/java/choice_types/DefaultEmptyChoiceTest.java
package choice_types;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import choice_types.default_empty_choice.DefaultEmptyChoice;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class DefaultEmptyChoiceTest
{
    @Test
    public void bitStreamReaderConstructor() throws IOException, ZserioError
    {
        final byte tag = VARIANT_B_SELECTOR;
        final short value = 234;
        final BitBuffer buffer = writeDefaultEmptyChoiceToBitBuffer(tag, value);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(buffer.getBuffer(),
                buffer.getBitSize());
        final DefaultEmptyChoice defaultEmptyChoice = new DefaultEmptyChoice(reader, tag);
        assertEquals(tag, defaultEmptyChoice.getTag());
        assertEquals((short)value, defaultEmptyChoice.getB());
    }

    @Test
    public void choiceTag()
    {
        DefaultEmptyChoice defaultEmptyChoice = new DefaultEmptyChoice(VARIANT_A_SELECTOR);
        assertEquals(DefaultEmptyChoice.CHOICE_a, defaultEmptyChoice.choiceTag());

        defaultEmptyChoice = new DefaultEmptyChoice(VARIANT_B_SELECTOR);
        assertEquals(DefaultEmptyChoice.CHOICE_b, defaultEmptyChoice.choiceTag());

        defaultEmptyChoice = new DefaultEmptyChoice(DEFAULT_SELECTOR);
        assertEquals(DefaultEmptyChoice.UNDEFINED_CHOICE, defaultEmptyChoice.choiceTag());
    }

    @Test
    public void bitStreamWriterWrite() throws IOException, ZserioError
    {
        final DefaultEmptyChoice defaultEmptyChoiceA = new DefaultEmptyChoice(VARIANT_A_SELECTOR);
        final byte byteValueA = 99;
        defaultEmptyChoiceA.setA(byteValueA);
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        defaultEmptyChoiceA.write(writer);
        ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray(),
                writer.getBitPosition());
        final DefaultEmptyChoice readDefaultEmptyChoiceA = new DefaultEmptyChoice(reader, VARIANT_A_SELECTOR);
        assertEquals(byteValueA, readDefaultEmptyChoiceA.getA());

        final DefaultEmptyChoice defaultEmptyChoiceB = new DefaultEmptyChoice(VARIANT_B_SELECTOR);
        final short shortValueB = 234;
        defaultEmptyChoiceB.setB(shortValueB);
        writer = new ByteArrayBitStreamWriter();
        defaultEmptyChoiceB.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final DefaultEmptyChoice readDefaultEmptyChoiceB = new DefaultEmptyChoice(reader, VARIANT_B_SELECTOR);
        assertEquals(shortValueB, readDefaultEmptyChoiceB.getB());

        final DefaultEmptyChoice defaultEmptyChoiceDefault = new DefaultEmptyChoice(DEFAULT_SELECTOR);
        writer = new ByteArrayBitStreamWriter();
        defaultEmptyChoiceDefault.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final DefaultEmptyChoice readDefaultEmptyChoiceDefault =
                new DefaultEmptyChoice(reader, DEFAULT_SELECTOR);
        assertEquals(DEFAULT_SELECTOR, readDefaultEmptyChoiceDefault.getTag());
    }

    private BitBuffer writeDefaultEmptyChoiceToBitBuffer(byte tag, short value) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            switch (tag)
            {
            case 1:
                writer.writeByte((byte)value);
                break;

            case 2:
                writer.writeShort(value);
                break;

            default:
                break;
            }

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private static byte VARIANT_A_SELECTOR = 1;
    private static byte VARIANT_B_SELECTOR = 2;
    private static byte DEFAULT_SELECTOR = 3;
}
